### PR TITLE
TEST: Refactoring integration tests from 0.1.x

### PR DESCRIPTION
**Pull request by bluemonk3y**
#340

### CODE REVIEW

File: ksql-engine/src/test/java/io/confluent/ksql/IntegrationTestHarness.java [hunk 1]
```diff
@@ -199,13 +211,13 @@ public static void main(String args[]) throws Exception {
...
     ksqlContext.sql("CREATE STREAM orders (ORDERTIME bigint, ORDERID varchar, ITEMID varchar, "
-                    + "ORDERUNITS double, PRICEARRAY array<double>, KEYVALUEMAP map<varchar, double>) WITH (kafka_topic='TestTopic', value_format='JSON');");
+            + "ORDERUNITS double, PRICEARRAY array<double>, KEYVALUEMAP map<varchar, double>) WITH (kafka_topic='TestTopic', value_format='JSON');");
     ksqlContext.sql("CREATE STREAM bigorders AS SELECT * FROM orders WHERE ORDERUNITS > 40;");
 
     Map<String, GenericRow> results = testHarness.consumeData("BIGORDERS",
-                                                                         orderDataProvider.schema
-                                                                             (), 4,
-                                                                         new StringDeserializer());
+            orderDataProvider.schema
```
Comment (dguy) [inline ksql-engine/src/test/java/io/confluent/ksql/IntegrationTestHarness.java]: for the line above, the `results` variable is never used some might aswell remove it?

File: ksql-engine/src/test/java/io/confluent/ksql/IntegrationTestHarness.java [hunk 2]
```diff
@@ -174,7 +186,7 @@ public void stop() {
   }
 
   public void assertExpectedResults(Map<String, GenericRow> actualResult,
-                                           Map<String, GenericRow> expectedResult) {
+                                    Map<String, GenericRow> expectedResult) {
```
Comment (dguy) [inline ksql-engine/src/test/java/io/confluent/ksql/IntegrationTestHarness.java]: I know not really part of this PR, but... this method can be replaced with `assertThat(actualResult, equalTo(expectedResult)`

File: ksql-engine/src/test/java/io/confluent/ksql/StreamsSelectAndProjectIntTest.java [hunk 3]
```diff
@@ -39,6 +42,45 @@ public void after() throws Exception {
...
+                    .timestamp() + 10000, "8", recordMetadataMap.get("8").timestamp() + 10000,
+            recordMetadataMap.get("8").timestamp
+                    () + 100, "ORDER_6", "ITEM_8")));
+
+    Schema resultSchema = ksqlContext.getMetaStore().getSource(stream2Name).getSchema();
+
+    Map<String, GenericRow> results1 = testHarness.consumeData(stream1Name, resultSchema , expectedResults.size(), new StringDeserializer());
+    Map<String, GenericRow> results2 = testHarness.consumeData(stream2Name, resultSchema , expectedResults.size(), new StringDeserializer());
+
+    System.out.println("Results:" + results1);
```
Comment (dguy) [inline ksql-engine/src/test/java/io/confluent/ksql/StreamsSelectAndProjectIntTest.java]: remove this?

File: ksql-engine/src/test/java/io/confluent/ksql/StreamsSelectAndProjectIntTest.java [hunk 4]
```diff
@@ -39,6 +42,45 @@ public void after() throws Exception {
...
+
+    Assert.assertEquals(expectedResults.size(), results2.size());
```
Comment (dguy) [inline ksql-engine/src/test/java/io/confluent/ksql/StreamsSelectAndProjectIntTest.java]: These 2 lines can be replaced with `assertThat(results2, expectedResults)`

File: ksql-engine/src/test/java/io/confluent/ksql/StreamsSelectAndProjectIntTest.java [hunk 5]
```diff
@@ -39,6 +42,45 @@ public void after() throws Exception {
...
+
+    Map<String, GenericRow> expectedResults = new HashMap<>();
+    expectedResults.put("8", new GenericRow(Arrays.asList(null, null, "8", recordMetadataMap.get("8")
+                    .timestamp() + 10000, "8", recordMetadataMap.get("8").timestamp() + 10000,
+            recordMetadataMap.get("8").timestamp
+                    () + 100, "ORDER_6", "ITEM_8")));
+
+    Schema resultSchema = ksqlContext.getMetaStore().getSource(stream2Name).getSchema();
+
+    Map<String, GenericRow> results1 = testHarness.consumeData(stream1Name, resultSchema , expectedResults.size(), new StringDeserializer());
```
Comment (dguy) [inline ksql-engine/src/test/java/io/confluent/ksql/StreamsSelectAndProjectIntTest.java]: Any reason why we do this twice? `results1` is only used for the sysout print

File: ksql-engine/src/test/java/io/confluent/ksql/StreamsSelectAndProjectIntTest.java [hunk 6]
```diff
@@ -49,15 +91,15 @@ public void testSelectProjectKeyTimestamp() throws Exception {
...
     ksqlContext.sql("CREATE STREAM PROJECT_KEY_TIMESTAMP AS SELECT ROWKEY AS RKEY, ROWTIME AS RTIME, ITEMID FROM ORDERS WHERE ORDERUNITS > 20 AND ITEMID = 'ITEM_8';");
 
-    Schema resultSchema = SchemaUtil.removeImplicitRowTimeRowKeyFromSchema(ksqlContext.getMetaStore().getSource("PROJECT_KEY_TIMESTAMP").getSchema());
+    Schema resultSchema = ksqlContext.getMetaStore().getSource("PROJECT_KEY_TIMESTAMP").getSchema();
 
     Map<String, GenericRow> results = testHarness.consumeData("PROJECT_KEY_TIMESTAMP", resultSchema , dataProvider.data().size(), new StringDeserializer());
 
     Map<String, GenericRow> expectedResults = new HashMap<>();
-    expectedResults.put("8", new GenericRow(Arrays.asList("8", recordMetadataMap.get("8").timestamp(), "ITEM_8")));
+    expectedResults.put("8", new GenericRow(Arrays.asList(null, null, "8", recordMetadataMap.get("8").timestamp(), "ITEM_8")));
```
Comment (dguy) [inline ksql-engine/src/test/java/io/confluent/ksql/StreamsSelectAndProjectIntTest.java]: nit: this could be `Collections.singletonMap(...)`

File: ksql-engine/src/test/java/io/confluent/ksql/StreamsSelectAndProjectIntTest.java [hunk 7]
```diff
@@ -49,15 +91,15 @@ public void testSelectProjectKeyTimestamp() throws Exception {
...
 
     Assert.assertEquals(expectedResults.size(), results.size());
-    assertExpectedResults(results, expectedResults);
+    assertExpectedResults(expectedResults, results);
```
Comment (dguy) [inline ksql-engine/src/test/java/io/confluent/ksql/StreamsSelectAndProjectIntTest.java]: as above `assertThat(...)`

File: ksql-engine/src/test/java/io/confluent/ksql/UdfIntTest.java [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+import java.util.concurrent.TimeoutException;
+
+import static io.confluent.ksql.util.MetaStoreFixture.assertExpectedResults;
+
+public class UdfIntTest {
+
+  private IntegrationTestHarness testHarness;
+  private KsqlContext ksqlContext;
+  private Map<String, RecordMetadata> recordMetadataMap;
+  String topicName = "TestTopic";
```
Comment (dguy) [inline ksql-engine/src/test/java/io/confluent/ksql/UdfIntTest.java]: `private` ?

File: ksql-engine/src/test/java/io/confluent/ksql/UdfIntTest.java [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+            "ITEMID, ORDERUNITS*10, PRICEARRAY[0]+10, KEYVALUEMAP['key1']*KEYVALUEMAP['key2']+10, PRICEARRAY[1]>1000",
+            "ORDERS",
+            "ORDERUNITS > 20 AND ITEMID LIKE '%_8'"
+    );
+
+    ksqlContext.sql(queryString);
+
+    Schema resultSchema = ksqlContext.getMetaStore().getSource(testStreamName).getSchema();
+
+    Map<String, GenericRow> expectedResults = new HashMap<>();
```
Comment (dguy) [inline ksql-engine/src/test/java/io/confluent/ksql/UdfIntTest.java]: `Collections.singletonMap(...)`? also elsewhere in this test class

File: ksql-engine/src/test/java/io/confluent/ksql/UdfIntTest.java [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+    expectedResults.put("8", new GenericRow(Arrays.asList(null, null, "ITEM_8", 800.0, 1110.0, 12.0, true)));
+
+    Map<String, GenericRow> results = testHarness.consumeData(testStreamName, resultSchema, 4, new StringDeserializer());
+
+    Assert.assertEquals(expectedResults.size(), results.size());
```
Comment (dguy) [inline ksql-engine/src/test/java/io/confluent/ksql/UdfIntTest.java]: as above and elsewhere in this test class

File: ksql-engine/src/test/java/io/confluent/ksql/UdfIntTest.java [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+            selectColumns,
+            "ORDERS",
+            "ORDERUNITS > 20 AND ITEMID LIKE '%_8'"
+    );
+
+    ksqlContext.sql(queryString);
+
+    Schema resultSchema = ksqlContext.getMetaStore().getSource(streamName).getSchema();
+
+    Map<String, GenericRow> expectedResults = new HashMap<>();
```
Comment (dguy) [inline ksql-engine/src/test/java/io/confluent/ksql/UdfIntTest.java]: ditto

File: ksql-engine/src/test/java/io/confluent/ksql/UdfIntTest.java [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+    expectedResults.put("8", new GenericRow(Arrays.asList(null, null, 80, "true", 8.0, "80.0")));
+
+    Map<String, GenericRow> results = testHarness.consumeData(streamName, resultSchema, 4, new StringDeserializer());
+
+
+    Assert.assertEquals(expectedResults.size(), results.size());
```
Comment (dguy) [inline ksql-engine/src/test/java/io/confluent/ksql/UdfIntTest.java]: ditto

File: ksql-engine/src/test/java/io/confluent/ksql/UdfIntTest.java [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+
+    Schema resultSchema = SchemaUtil
+            .removeImplicitRowTimeRowKeyFromSchema(ksqlContext.getMetaStore().getSource(stream2Name).getSchema());
+
+    Map<String, GenericRow> expectedResults = new HashMap<>();
+    expectedResults.put("8", new GenericRow(Arrays.asList("8", recordMetadataMap.get("8").timestamp() +
+                    10000, "8", recordMetadataMap.get("8").timestamp() + 10000,
+            recordMetadataMap.get("8").timestamp() + 100, "ORDER_6", "ITEM_8")));
+
+    Map<String, GenericRow> results1 = testHarness.consumeData(stream1Name, resultSchema,expectedResults.size(), new StringDeserializer());
```
Comment (dguy) [inline ksql-engine/src/test/java/io/confluent/ksql/UdfIntTest.java]: do we need this one? it isn't used

File: ksql-engine/src/test/java/io/confluent/ksql/WindowingIntTest.java [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+    ksqlContext.sql(queryString);
+
+    Schema resultSchema = ksqlContext.getMetaStore().getSource(streamName).getSchema();
+
+
+    long firstItem8Window  = recordMetadataMap.get("8").timestamp() / windowSizeMilliseconds;
+    long secondItem8Window =   secondSetOfRecords.get("8").timestamp() / windowSizeMilliseconds;
+
+    Map<Windowed<String>, GenericRow> expectedResults = new HashMap<>();
+    if (firstItem8Window == secondItem8Window) {
```
Comment (dguy) [inline ksql-engine/src/test/java/io/confluent/ksql/WindowingIntTest.java]: is there a way we can make this always the case? It seems a bit odd that the test could have a different outcome each time it runs.

File: ksql-engine/src/test/java/io/confluent/ksql/WindowingIntTest.java [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+      expectedResults.put(
+              new Windowed<>("ITEM_8",new TimeWindow(0, 1)),
+              new GenericRow(Arrays.asList(null, null, "ITEM_8", 2, 160.0, 80.0, 2220.0))
+      );
+    }
+
+    Map<Windowed<String>, GenericRow> results = testHarness.consumeData(streamName, resultSchema, expectedResults.size(), new WindowedDeserializer<>(new StringDeserializer()));
+
+    System.out.println("Results:" + results + " size:" + results.size());
```
Comment (dguy) [inline ksql-engine/src/test/java/io/confluent/ksql/WindowingIntTest.java]: probably don't need this

File: ksql-engine/src/test/java/io/confluent/ksql/WindowingIntTest.java [hunk 16]
```diff
@@ -0,0 +1,103 @@
...
+
+    Assert.assertEquals(expectedResults.size(), results.size());
+    assertExpectedWindowedResults(results, expectedResults);
+  }
+
+  private void createOrdersStream() throws Exception {
+    ksqlContext.sql("CREATE STREAM ORDERS (ORDERTIME bigint, ORDERID varchar, ITEMID varchar, ORDERUNITS double, PRICEARRAY array<double>, KEYVALUEMAP map<varchar, double>) WITH (kafka_topic='TestTopic', value_format='JSON', key='ordertime');");
+  }
+
+  private OrderDataProvider publishOrdersTopicData() throws InterruptedException, TimeoutException, ExecutionException {
```
Comment (dguy) [inline ksql-engine/src/test/java/io/confluent/ksql/WindowingIntTest.java]: There are similar methods to this in other test classes. Should we move it to the `IntegrationTestHarness` class? Have the method take a `TestDataProvider` param and return the `recordMetadataMap` ?

**Comment by bluemonk3y** [comment]
@dguy  - think Ive covered most stuff. Moving the window test into another PR
